firebase-firestore/src/main/java/com/google/firebase/firestore/remote/RemoteStore.java

```diff
@@ -406,8 +430,9 @@ private void startWatchStream() {
         shouldStartWatchStream(),
         "startWatchStream() called when shouldStartWatchStream() is false.");
     watchChangeAggregator = new WatchChangeAggregator(this);
-    watchStream.start();
 
+    initializeSslIfNeeded();
```

Comment:
I'm not a fan of blocking here. Firestore is an event loop and we should reformulate this in terms of an event we need to handle. This avoids exposes threading constructs into code that currently is happily single threaded. Specifically, once SSL initialization is complete, we post that completion into the async queue.

The question then becomes, how to avoid touching the network before this is ready? I think is easy because the RemoteStore already understands how to run with the network is disabled. `canUseNetwork()` already gates every action that can touch the network. Our goal should be to piggyback off of that.

In particular, I think you should add a boolean to RemoteStore called `networkInitialized` or something like that. Default it to false, and the SslProvider completion should flip it to true in something like `handleNetworkInitialized()`.

Then factor out everything but `networkEnabled = true;` from `enableNetwork` into a helper (`maybeStartNetwork`?) and call it from `handleNetworkInitialized` as well.

Change `canUseNetwork` to check `networkInitialized && networkEnabled`.

Note that the simpler approach of just delaying `RemoteStore.start` until the network is initialized doesn't quite work, because the first action the user takes could be to disable the network. We can't have the completion of SSL initialization overriding that.

Reply:
I tried to figure out how to prevent blocking the thread, but my solution would have required some large changes in other parts of the SDK. The idea was basically that the client would no longer rely on the RemoteStore's ability to always be in a state that it can start the network from. 
The missing piece was that this doesn't have to be the case, since the RemoteStore can already catch up to the state it is supposed to be in (via `fillWritePipeline()`). Long story short - what you are suggesting here is a huge improvement and I wish I had come up with that myself. Thanks :)